### PR TITLE
tests: report mgas/s metric in evm benchmarks

### PR DESCRIPTION
**Pull request by shekhirin**
I think it's good to have `mgas/s` metric in EVM benchmarks, so anyone could compare geth's implementation with e.g. [evmone](https://github.com/ethereum/evmone)'s which has [the same reporting format](https://github.com/ethereum/evmone/blob/4ab4b4b928ed85fa6eea58cc135f31ab1f1ba8fa/test/bench/helpers.hpp#L145).

Ideally, we would be able to have all EVM implementations benchmarked against one metric that's comparable regardless of implementation language.

### CODE REVIEW

**Review by MariusVanDerWijden — APPROVED** [review]
SGTM
Looks like you updated the testdata submodule, could you please revert that change. It would be better imo to merge that separately

**Comment by shekhirin** [comment]
@MariusVanDerWijden agree, fixed!

File: tests/state_test.go [hunk 1]
```diff
@@ -242,12 +243,26 @@ func runBenchmark(b *testing.B, t *StateTest) {
...
 				statedb.RevertToSnapshot(snapshot)
+
+				elapsed := uint64(time.Since(start))
+				if elapsed < 1 {
+					elapsed = 1
+				}
+				gasUsed := msg.Gas() - leftOverGas
+				// Keep it as uint64, multiply 100 to get two digit float later
+				mgasps := (100 * 1000 * gasUsed) / elapsed
+				b.ReportMetric(float64(mgasps)/100, "mgas/s")
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply tests/state_test.go]: yep, agree. Fixed.

File: tests/state_test.go [hunk 2]
```diff
@@ -238,17 +239,36 @@ func runBenchmark(b *testing.B, t *StateTest) {
...
+				gasUsed += msg.Gas() - leftOverGas
+
 				statedb.RevertToSnapshot(snapshot)
 			}
+
+			if elapsed < 1 {
+				elapsed = 1
+			}
+			// Keep it as uint64, multiply 100 to get two digit float later
+			mgasps := (100 * 1000 * gasUsed) / elapsed
```
Comment (holiman) [inline tests/state_test.go:270]: Thought: could either the aggregation of gasUsed go above uint64Max, or, alternatively, 100*1000* mult here might push it over ? 
If so, we would be better off keeping track of it as a big.Int.

Comment (shekhirin) [reply tests/state_test.go:270]: it seems fine:
```bash
➜  go-ethereum git:(benchmark-evm-mgas) fd '.json' tests/evm-benchmarks/benchmarks | # Get all test file paths
xargs jq -r 'values[] | .env.currentGasLimit' | # Extract gas limit
xargs -L 1 cast --to-dec | # Hex -> dec
sort -u | # Unique 
awk '{ print 18446744073709551615 - 100 * 1000 * $0 * 30000 }' # math.MaxUint64 - 100 * 1000 * gasUsed * b.N

15446744073709551616
```

File: tests/state_test.go [hunk 3]
```diff
@@ -238,17 +239,36 @@ func runBenchmark(b *testing.B, t *StateTest) {
...
+				// Execute the message.
+				_, leftOverGas, err := evm.Call(sender, *msg.To(), msg.Data(), msg.Gas(), msg.Value())
 				if err != nil {
 					b.Error(err)
 					return
 				}
+
+				b.StopTimer()
+				elapsed += uint64(time.Since(start))
+				gasUsed += msg.Gas() - leftOverGas
```
Comment (holiman) [inline tests/state_test.go:261]: This is not quite accurate metric. We should also take into account the refunds when calculating the mgas/s. 
This error is also present in `core/blockchain_insert.go`, 
```
			"elapsed", common.PrettyDuration(elapsed), "mgasps", float64(st.usedGas) * 1000 / float64(elapsed),
```
... but we should fix it in both places, rather than do it erroneously here too.

**Review by holiman — CHANGES_REQUESTED** [review]
We should also take into account the refunds when calculating the mgas/s.

**Comment by holiman** [comment]
After looking into it a bit more, I have fixed the refunds in the benchmarks now. As for 

> This error is also present in core/blockchain_insert.go,
> ... but we should fix it in both places, rather than do it erroneously here too.

I no longer believe that to be the case. But the flow is a bit hard to follow, so I need to verify that experimentally. If it needs to be fixed, it should be fixed in a separate PR, since it touches some core stuff.

**Review by holiman — APPROVED** [review]
LGTM
